Autocorrection with `literal_expression_end_indentation` results in uncompilable code

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Using `swiftlint autocorrect` with `literal_expression_end_indentation` in complex expression results in uncompilable code.
Might be related to https://github.com/realm/SwiftLint/issues/2686

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths 
Linting 'ViewController.swift' (1/2)
Linting 'AppDelegate.swift' (2/2)
/<path redacted>/ViewController.swift:12:9: warning: Literal Expression End Indentation Violation: Array and dictionary literal end should have the same indentation as the line that started it. Expected 4, got 8. (literal_expression_end_indentation)
Done linting! Found 1 violation, 0 serious in 2 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? `0.34.0`
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Homebrew
* Paste your configuration file:

```yml
opt_in_rules:
  - literal_expression_end_indentation

```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcode-select -p`)? Version 10.2.1 (10E1001)
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

Run `swift autocorrect` on this code:

```swift
import UIKit

class ViewController: UIViewController {
  struct SomeStuct {
    let setup: (Int) -> Void
  }

  private func build() -> [SomeStuct] {
    return [
      SomeStuct(setup: { val in
        print("val", val)
      })]
  }

}
```

`swiftlint autocorrect` output:
```
$ swiftlint autocorrect
Loading configuration from '.swiftlint.yml'
Correcting Swift files at paths 
Correcting 'ViewController.swift' (1/2)
Correcting 'AppDelegate.swift' (2/2)
/<path redacted>/ViewController.swift:12:1 Corrected Literal Expression End Indentation
/<path redacted>/ViewController.swift:13:1 Corrected Literal Expression End Indentation
Done correcting 2 files!
```

Resulting code:
```
import UIKit

class ViewController: UIViewController {
  struct SomeStuct {
    let setup: (Int) -> Void
  }

  private func build() -> [SomeStuct] {
    return [
      SomeStuct(setup: { val in
        print("val", val)
    ] // compile error: Expected expression
    } // compile error: Expected ']' in container literal expression

} //compile error: Expected ')' in expression list
// compile error: Expected '}' in class
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -42,6 +42,14 @@
 
 ### Bug Fixes
 
+* Fix `literal_expression_end_indentation` autocorrection deleting source code
+  when the closing bracket of a multiline literal shares a line with the end of
+  a multiline last element (e.g. `...))]`). The corrector assumed everything
+  before the bracket on that line was indentation and replaced it; it now moves
+  only the bracket to its own line at the expected indentation.  
+  [Luan Câmara](https://github.com/luancamara)
+  [#2823](https://github.com/realm/SwiftLint/issues/2823)
+
 * Don't rewrite the type operand of an `is` / `as?` / `as!` cast (such as
   `x is A`) to `Self` in `prefer_self_in_static_references` when inside a
   class-like scope. `Self` is the dynamic type, so the rewrite silently changed

--- a/Source/SwiftLintBuiltInRules/Rules/Style/LiteralExpressionEndIndentationRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/LiteralExpressionEndIndentationRule.swift
@@ -69,6 +69,12 @@ struct LiteralExpressionEndIndentationRule: Rule, OptInRule {
                key: value
                ↓]
             """),
+            Example("""
+            let x = [
+               foo(
+                  1,
+                  2)↓]
+            """),
         ],
         corrections: [
             Example("""
@@ -117,6 +123,18 @@ struct LiteralExpressionEndIndentationRule: Rule, OptInRule {
             ] + [
                3,
                4
+            ]
+            """),
+            Example("""
+            let x = [
+               foo(
+                  1,
+                  2)↓]
+            """): Example("""
+            let x = [
+               foo(
+                  1,
+                  2)
             ]
             """),
         ]
@@ -176,7 +194,11 @@ extension LiteralExpressionEndIndentationRule: CorrectableRule {
         }
 
         let correction = contents.substring(from: expected.location, length: expected.length)
-        contents = contents.replacingCharacters(in: actualIndices, with: correction)
+        if contents[actualIndices].allSatisfy(\.isWhitespace) {
+            contents = contents.replacingCharacters(in: actualIndices, with: correction)
+        } else {
+            contents.insert(contentsOf: "\n" + correction, at: actualIndices.upperBound)
+        }
 
         return true
     }

